includes/class-wc-order-item.php

```diff
@@ -0,0 +1,574 @@
+		if ( ! $this->get_id() ) {
+			$this->create();
+		} else {
+			$this->update();
+		}
+		$this->save_meta_data();
+
+		return $this->get_id();
+	}
+
+	/**
+	 * Delete data from the database.
+	 * @since 2.7.0
+	 */
+	public function delete() {
+		if ( $this->get_id() ) {
+			global $wpdb;
+			do_action( 'woocommerce_before_delete_order_item', $this->get_id() );
+			$wpdb->delete( $wpdb->prefix . 'woocommerce_order_items', array( 'order_item_id' => $this->get_id() ) );
+			$wpdb->delete( $wpdb->prefix . 'woocommerce_order_itemmeta', array( 'order_item_id' => $this->get_id() ) );
+			do_action( 'woocommerce_delete_order_item', $this->get_id() );
+		}
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Meta Data Handling
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Get All Meta Data
+	 * @return array
+	 */
+	public function get_meta_data() {
+		return $this->_meta_data;
+	}
+
+	/**
+	 * Expands things like term slugs before return.
+	 * @param string $hideprefix (default: _)
+	 * @return array
+	 */
+	public function get_formatted_meta_data( $hideprefix = '_' ) {
+		$formatted_meta = array();
+		$meta_data      = $this->get_meta_data();
+
+		foreach ( $meta_data as $meta_id => $meta ) {
+			if ( "" === $meta->value || is_serialized( $meta->value ) || ( ! empty( $hideprefix ) && substr( $meta->key, 0, 1 ) === $hideprefix ) ) {
+				continue;
+			}
+
+			$attribute_key = urldecode( str_replace( 'attribute_', '', $meta->key ) );
+			$display_key   = wc_attribute_label( $attribute_key, is_callable( array( $this, 'get_product' ) ) ? $this->get_product() : false );
+			$display_value = $meta->value;
+
+			if ( taxonomy_exists( $attribute_key ) ) {
+				$term = get_term_by( 'slug', $meta->value, $attribute_key );
+				if ( ! is_wp_error( $term ) && is_object( $term ) && $term->name ) {
+					$display_value = $term->name;
+				}
+			}
+
+			$formatted_meta[ $meta_id ] = (object) array(
+				'key'           => $meta->key,
+				'value'         => $meta->key,
+				'display_key'   => apply_filters( 'woocommerce_order_item_display_meta_key', $display_key ),
+				'display_value' => apply_filters( 'woocommerce_order_item_display_meta_value', $display_value ),
+			);
+		}
+
+		return $formatted_meta;
+	}
+
+	/**
+	 * Internal meta keys we don't want exposed as part of meta_data.
+	 * @return array()
+	 */
+	protected function get_internal_meta_keys() {
+		return array();
+	}
+
+	/**
+	 * Get Meta Data by Key
+	 * @param string $key
+	 * @param bool $single return first found meta with key, or all with $key
+	 * @return mixed
+	 */
+	public function get_meta( $key = '', $single = true ) {
+		$meta_ids = array_keys( wp_list_pluck( $this->_meta_data, 'key' ), $key );
+		$value    = '';
+
+		if ( $meta_ids ) {
+			if ( $single ) {
+				$value   = $this->_meta_data[ current( $meta_ids ) ]->value;
+			} else {
+				$value = array_intersect_key( $this->_meta_data, $meta_ids );
```

Comment:
I'm fixing it in the new branch/class, but wanted to note that $meta_ids needs an array_flip here to work properly. The IDs are in values at this point.